src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -26,6 +27,36 @@ public sealed partial class NpgsqlReadBuffer
 
         internal Stream Underlying { private get; set; }
 
+        readonly Socket _underlyingSocket;
+
+        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+
+        /// <summary>
+        /// Underlying socket ReceiveTimeout
+        /// </summary>
+        internal TimeSpan Timeout
+        {
+            get => _currentSocketTimeout;
+            set
+            {
+                if (_currentSocketTimeout != value)
+                {
+                    if (_underlyingSocket == null)
```

Comment:
This seems like it could only be true if there's a bug in Npgsql (timeout should never be set when there's no socket, i.e. when pregenerating messages). If so, I think this should be a Debug.Assert rather than a runtime check.